# Data Handling

## EDA Tiller System

### Loading file

Download the lab data by following this [link](https://drive.google.com/file/d/1X5uXVFrl7SuMUKe7CJaiCmGEOhokorh2/view?usp=sharing)

a) Load the Tiller_order_data.csv file into a DataFrame

In [ ]:
import pandas as pd

df = pd.read_csv("Tiller_order_data.csv")

/var/folders/zx/5hkdnl813p716fh3kqzd3wm40000gn/T/ipykernel_82714/2575152319.py:3: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("Tiller_order_data.csv")


### Data Discovery

b) Display the first 5 rows of this dataframe

In [ ]:
df.head()

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device              date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0  2019-01-12 13:02:17 UTC   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0  2019-01-16 19:39:09 UTC   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0  2019-01-12 14:18:46 UTC   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0  2019-01-24 17:49:12 UTC   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0  2019-01-12 15:25:06 UTC   

               date_closed                    dim_name dim_status  dim_type  \
0  2019-01-12 19:58:38 UTC                     vincent     CLOSED         1   
1  2019-01-16 22:10:50 UTC  frere et soeur avec pierre     CLOSED         1   
2  2019-01-12 19:50:32 UTC                      rachel     CLOSED         1   
3  2019-01-24 21:58:59 UTC                  Groupe PEL     CLOSED         1   
4  2019-01-12 19:21:03 UTC                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  
0         NaN   Tiller iPAD              1            45.5            45.5  
1         NaN   Tiller iPAD              2            49.8            49.8  
2         NaN   Tiller iPAD              1            27.4            27.4  
3         NaN  tiller-order              3            60.0            60.0  
4         NaN   Tiller iPAD              2            39.9            39.9

c) What does a row of this dataframe represent? What information can you find?

One line corresponds to one command.
You can find for example:
- the brand associated with this order
- the opening and closing time of the order
- The number of customers associated with this order
- The amount of the order
- etc.

d) How many orders are there in this dataframe? How many columns?

In [ ]:
df.shape

(117304, 17)

#### Column selection

e) Select only the id_order, id_store, date_opened and date_closed columns of this DataFrame and save the result in a variable df_sub

In [ ]:
df_sub = df[["id_order", "id_store", "date_opened", "date_closed"]]
df_sub

id_order  id_store              date_opened              date_closed
0       55538867      8052  2019-01-12 13:02:17 UTC  2019-01-12 19:58:38 UTC
1       56035309      8052  2019-01-16 19:39:09 UTC  2019-01-16 22:10:50 UTC
2       55550051      8052  2019-01-12 14:18:46 UTC  2019-01-12 19:50:32 UTC
3       57000119      8052  2019-01-24 17:49:12 UTC  2019-01-24 21:58:59 UTC
4       55558817      8052  2019-01-12 15:25:06 UTC  2019-01-12 19:21:03 UTC
...          ...       ...                      ...                      ...
117299  61630471      4542  2019-02-28 13:13:21 UTC  2019-02-28 13:38:45 UTC
117300  76772730      4542  2019-06-11 11:08:01 UTC  2019-06-11 11:10:26 UTC
117301  77799227      4542  2019-06-17 10:30:50 UTC  2019-06-17 10:32:45 UTC
117302  60627160      4542  2019-02-21 11:50:58 UTC  2019-02-21 11:51:56 UTC
117303  93223821      4542  2019-09-18 11:32:35 UTC  2019-09-18 11:39:38 UTC

[117304 rows x 4 columns]

#### Row selection

f) Filter df_sub only on lines concerning id_store 4542. How many orders have there been in this restaurant?

In [ ]:
df_sub[df_sub["id_store"]==4542]

id_order  id_store              date_opened              date_closed
99663    82814018      4542  2019-07-16 10:57:10 UTC  2019-07-16 10:58:58 UTC
99664    57084271      4542  2019-01-25 11:49:43 UTC  2019-01-25 11:53:03 UTC
99665   132161083      4542  2020-07-10 11:35:51 UTC  2020-07-10 11:35:53 UTC
99666   109101613      4542  2019-12-17 11:35:02 UTC  2019-12-17 11:35:26 UTC
99667   133761191      4542  2020-07-20 11:01:46 UTC  2020-07-20 11:02:28 UTC
...           ...       ...                      ...                      ...
117299   61630471      4542  2019-02-28 13:13:21 UTC  2019-02-28 13:38:45 UTC
117300   76772730      4542  2019-06-11 11:08:01 UTC  2019-06-11 11:10:26 UTC
117301   77799227      4542  2019-06-17 10:30:50 UTC  2019-06-17 10:32:45 UTC
117302   60627160      4542  2019-02-21 11:50:58 UTC  2019-02-21 11:51:56 UTC
117303   93223821      4542  2019-09-18 11:32:35 UTC  2019-09-18 11:39:38 UTC

[17641 rows x 4 columns]

### Data analysis

g) Going back to the initial DataFrame, what are the different possible statuses of an order? How many orders are there per status?

In [ ]:
print(df["dim_status"].unique())

print(df.groupby(["dim_status"], as_index=False).size())

['CLOSED' 'IN_PROGRESS']
    dim_status    size
0       CLOSED  116939
1  IN_PROGRESS     365


h) What is the proportion of orders whose status is CLOSE?

In [ ]:
mask_order_close = df["dim_status"] == "CLOSED"

print(df[mask_order_close].shape[0] / df.shape[0])

0.9968884266521175


i) How many restaurants are there in this data? What are their IDs?

In [ ]:
stores = df["id_store"].unique()

print(stores)
print(len(stores))

[8052 8347 8283 9084 4337  360 1796 7965 6293 4542]
10


j) Calculate the turnover per restaurant. Which restaurant has the highest turnover?

In [ ]:
df.groupby(["id_store"])["m_cached_payed"].sum().sort_values(ascending=False)

id_store
7965    1138690.95
4337     865357.89
9084     479019.75
360      387089.14
8283     369822.82
6293     359471.00
8052     327758.85
1796     321099.69
4542     243728.69
8347         23.00
Name: m_cached_payed, dtype: float64

k) Count the number of orders per restaurant. Which restaurant had the most orders in this data? Is it the same as the one with the highest turnover?

In [ ]:
df.groupby(["id_store"]).size().sort_values(ascending=False)

id_store
1796    27349
4337    20003
4542    17641
6293    16103
8283    10023
7965     9091
9084     7804
8052     4906
360      4383
8347        1
dtype: int64

l) Calculate the average price of an order for each restaurant. Do you think you face the same types of restaurants?

In [ ]:
df_ca = df.groupby(["id_store"])["m_cached_price"].agg(['size', 'sum']).reset_index()
df_ca["avg_order_price"] = df_ca["sum"] / df_ca["size"]
df_ca.sort_values("avg_order_price", ascending=False)

id_store   size         sum  avg_order_price
5      7965   9091  1138585.95       125.243202
0       360   4383   386934.14        88.280662
6      8052   4906   327758.85        66.807756
9      9084   7804   479040.75        61.384002
2      4337  20003   865847.30        43.285872
7      8283  10023   369822.82        36.897418
8      8347      1       23.00        23.000000
4      6293  16103   359718.00        22.338570
3      4542  17641   243789.69        13.819494
1      1796  27349   321099.69        11.740820

m) The problem is that an order can be placed for several people. We rather want to know the price spent on average by a person in the restaurant.

From the initial dataframe, add a column representing the price of the order per customer

In [ ]:
df["price_per_customer"] = df["m_cached_payed"] / df["m_nb_customer"]
df.head()

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device              date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0  2019-01-12 13:02:17 UTC   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0  2019-01-16 19:39:09 UTC   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0  2019-01-12 14:18:46 UTC   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0  2019-01-24 17:49:12 UTC   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0  2019-01-12 15:25:06 UTC   

               date_closed                    dim_name dim_status  dim_type  \
0  2019-01-12 19:58:38 UTC                     vincent     CLOSED         1   
1  2019-01-16 22:10:50 UTC  frere et soeur avec pierre     CLOSED         1   
2  2019-01-12 19:50:32 UTC                      rachel     CLOSED         1   
3  2019-01-24 21:58:59 UTC                  Groupe PEL     CLOSED         1   
4  2019-01-12 19:21:03 UTC                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
0         NaN   Tiller iPAD              1            45.5            45.5   
1         NaN   Tiller iPAD              2            49.8            49.8   
2         NaN   Tiller iPAD              1            27.4            27.4   
3         NaN  tiller-order              3            60.0            60.0   
4         NaN   Tiller iPAD              2            39.9            39.9   

   price_per_customer  
0               45.50  
1               24.90  
2               27.40  
3               20.00  
4               19.95

n) Using the previous column, calculate the average price spent by each person in each of the restaurants.

In [ ]:
df.groupby(["id_store"])["price_per_customer"].mean()

id_store
360           NaN
1796          NaN
4337          NaN
4542    13.816036
6293          NaN
7965          NaN
8052    16.698658
8283    36.897418
8347    11.500000
9084          NaN
Name: price_per_customer, dtype: float64

o) It seems that there are data problems because for some restaurants the average price is lower ...

Do you have any idea why?

Replace the values ​​at 0 in the m_nb_customer column with 1 then delete the rows whose m_cached_price column is at 0

In [ ]:
df.loc[df.m_nb_customer == 0, "m_nb_customer"] = 1
df = df[df["m_cached_payed"] > 0]


p) Recalculate the price_per_customer column and the average price spent by each person in a restaurant

In [ ]:
df["price_per_customer"] = df["m_cached_payed"] / df["m_nb_customer"]

df.groupby(["id_store"])["price_per_customer"].mean()

id_store
360      32.081429
1796     11.832542
4337     43.353690
4542     13.819953
6293     16.806272
7965    102.439694
8052     16.722518
8283     36.901100
8347     11.500000
9084     30.571901
Name: price_per_customer, dtype: float64

### Temporal analysis

Restaurant 7965 has a higher price per person than the rest. Let's look at the time spent at the table to find out if this can correspond to a large restaurant.

q) Convert date_opened and date_closed columns to datetime format

In [ ]:
df.date_closed = pd.to_datetime(df.date_closed)
df.date_opened = pd.to_datetime(df.date_opened)

r) Run the code below to calculate the time spent between the open and close date of an order in minutes.

In [ ]:
df["order_duration"] = (df.date_closed - df.date_opened).dt.total_seconds() / 60


s) Calculate the average time at the table for each of the restaurants.

In [ ]:
df.groupby(["id_store"])["order_duration"].mean()

id_store
360      96.262090
1796      3.399033
4337      6.251355
4542      1.893868
6293     29.594241
7965     53.715114
8052    164.739022
8283      2.005228
8347      3.900000
9084    103.025325
Name: order_duration, dtype: float64

It seems that there are no fast food type restaurants! Open order durations are relatively short.

### Join operations

t) Download data from [order_line](https://drive.google.com/file/d/1G7Jbq_m-tJrHRhL-44j8XFOVs_zPil_9/view?usp=sharing) and [store_data](https://drive.google.com/file/d/1W032PaySL7gouyX3bEKgndtm6qVta7Qp/view?usp=sharing). Load them in Python.

In [ ]:
order_line = pd.read_csv("Tiller_order_line.csv")
store_data = pd.read_csv("Tiller_store_data.csv")

u) What new information do you have access to thanks to this data?

In [ ]:
order_line.head()

id_order_line   id_order             date_opended             date_created  \
0      626609716  139058632  2020-08-25 17:27:14 UTC  2020-08-25 17:27:25 UTC   
1      403167240   94770355  2019-09-27 07:47:14 UTC  2019-09-27 07:47:18 UTC   
2      578564178  127794604  2020-06-12 11:07:09 UTC  2020-06-12 11:07:25 UTC   
3      329942564   77841193  2019-06-17 13:13:30 UTC  2019-06-17 13:13:43 UTC   
4      348375251   82283609  2019-07-12 20:14:51 UTC  2019-07-12 20:15:38 UTC   

   m_quantity  m_unit_price  m_unit_price_exc_vat  m_total_price_inc_vat  \
0         1.0         15.00             12.500000                  15.00   
1         1.0         25.90             21.583333                  25.90   
2         1.0          4.20              3.818182                   4.20   
3         1.0          2.20              2.085308                   2.20   
4         1.0         -5.16             -5.160000                  -5.16   

   m_total_price_exc_vat  m_tax_percent  m_discount_amount  dim_type  \
0              12.500000          0.200                0.0   PRODUCT   
1              21.583333          0.200                0.0   PRODUCT   
2               3.818182          0.100                0.0   PRODUCT   
3               2.085308          0.055                0.0   PRODUCT   
4              -5.160000          0.000                0.0  DISCOUNT   

           dim_category                              dim_name dim_status  \
0           VINS ROUGES  TWIST AGAIN 2018 VDF - YOANN ROMAGNE       DONE   
1  LIQUIDES 50ML AUTRES                            MOMMY CAKE       DONE   
2              Déjeuner                      FORMULE JUS 30CL       DONE   
3              BOISSONS                 EVIAN - BADOIT - COCA       DONE   
4               VARIOUS                              DISCOUNT       DONE   

   m_nb_payed dim_feature_type dim_unit_measure dim_unit_measure_display  
0           0              NaN             UNIT                     UNIT  
1           0              NaN             UNIT                     UNIT  
2           0              NaN             UNIT                     UNIT  
3           0              NaN             UNIT                     UNIT  
4           0              NaN              NaN                      NaN

In [ ]:
store_data.head()

id_store               date_created dim_zipcode dim_country dim_currency
0      2576  2016-07-12 16:15:31+00:00       34120          FR          EUR
1      2538  2016-07-07 16:35:44+00:00       47800          FR          EUR
2      2468  2016-06-30 12:44:25+00:00       06470          FR          EUR
3      2375  2016-06-21 12:52:36+00:00       76000          FR          EUR
4      2328  2016-06-16 13:59:52+00:00       75018          FR          EUR

- Order line: The different products making up an order
- Store data: Information about the restaurant/bar

v) How many orders were placed in the 2nd arrondissement of Paris? It will take merging store_data with the seed data to answer it.

In [ ]:
order_data_w_store_info = df.merge(store_data, on="id_store")

mask = order_data_w_store_info.dim_zipcode == "75002"
order_data_w_store_info[mask].shape[0]

59357

w) Take restaurant 1796. Its average order time was very short.

Show the most ordered items in this restaurant. To do this keep only the id_store and id_order columns of order_data and the id_order, id_order_line and dim_name columns of order_line.

Be careful, the dim_name column exists both in order_data and in order_line.

In [ ]:
df_all = df[["id_order", "id_store"]].merge(order_line[["id_order", "id_order_line", "dim_name", "dim_category"]], on="id_order")

df_all[df_all.id_store == 1796].groupby(["dim_name"]).size().sort_values(ascending=False).head(30)

dim_name
COMPOSITION                 10152
+1€                          6244
FORMULE DESSERT              4091
FORMULE JUS 30CL             4082
FORMULE JUS 30CL / SOUPE     2534
CHICKEN AVOCADO              1553
POULET CURRY                 1492
SAUMON MI CUIT               1377
AGNEAU AUX EPICES            1241
JUS 30CL                     1221
SWEET KISS                   1150
FORMULE 40CL                 1035
MINTY WAY                    1026
BURRATA                      1010
JUS 40CL                     1002
+0,5€                         918
FORMULE DESSERT/BOISSON       838
+ 1€                          801
MENU 30cL                     783
BOL                           736
AGNEAU MIJOTE                 734
TOFU COCO                     729
KUBRICK                       660
ANTIDOTE                      658
VEGGIBALL                     623
KIWI KALE                     609
POKE SAUMON                   577
CHOP SUEY                     573
EVIAN - BADOIT - COCA         568
DISCO

Is it a fast lunch restaurant in your opinion?

### Data cleaning

x) [More difficult] The dim_name column of order_data data is unstructured because it is a free field. We want to clean it by creating a new "processed_dim_name" column with:

- Any value composed solely of numbers, whether or not preceded by a capital letter (example: B124, A7, 42) will be categorized as "Code"
- Any value including "LaFourchette" will be categorized as "LaFourchette"
- All remaining values ​​will be categorized as "Personne"
- Empty values ​​or character strings containing only spaces will be put empty

You can look at the `pd.isna()` function, the regex with the `re` library.

In [ ]:
import re
import numpy as np

processed_dim_name_list = []
for name in df.dim_name.values:
    processed_dim_name = None
    if pd.isna(name):
        processed_dim_name = name
    elif re.match("[A-Z]?\d+", name):
        processed_dim_name = "Code"
    elif "LaFourchette" in name:
        processed_dim_name = "LaFourchette"
    elif name.rstrip() == "":
        processed_dim_name = np.nan
    else:
        processed_dim_name = "Personne"
    processed_dim_name_list.append(processed_dim_name)

df["processed_dim_name"] = processed_dim_name_list


In [ ]:
df.head()

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device               date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0 2019-01-12 13:02:17+00:00   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0 2019-01-16 19:39:09+00:00   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0 2019-01-12 14:18:46+00:00   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0 2019-01-24 17:49:12+00:00   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0 2019-01-12 15:25:06+00:00   

                date_closed                    dim_name dim_status  dim_type  \
0 2019-01-12 19:58:38+00:00                     vincent     CLOSED         1   
1 2019-01-16 22:10:50+00:00  frere et soeur avec pierre     CLOSED         1   
2 2019-01-12 19:50:32+00:00                      rachel     CLOSED         1   
3 2019-01-24 21:58:59+00:00                  Groupe PEL     CLOSED         1   
4 2019-01-12 19:21:03+00:00                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
0         NaN   Tiller iPAD              1            45.5            45.5   
1         NaN   Tiller iPAD              2            49.8            49.8   
2         NaN   Tiller iPAD              1            27.4            27.4   
3         NaN  tiller-order              3            60.0            60.0   
4         NaN   Tiller iPAD              2            39.9            39.9   

   price_per_customer  order_duration processed_dim_name  
0               45.50      416.350000           Personne  
1               24.90      151.683333           Personne  
2               27.40      331.766667           Personne  
3               20.00      249.783333           Personne  
4               19.95      235.950000           Personne

y) To meet the needs of a team, we want to calculate the number of orders per restaurant and processed_dim_name. Create this dataframe.

In [ ]:
df_processed_dim_name = df.groupby(["id_store", "processed_dim_name"], as_index=False).size()
df_processed_dim_name

id_store processed_dim_name   size
0        360               Code    529
1        360       LaFourchette      2
2        360           Personne      2
3       1796               Code  24518
4       4337           Personne    145
5       4542               Code   9145
6       6293               Code    240
7       6293           Personne    271
8       7965               Code    725
9       7965           Personne   1909
10      8052               Code    495
11      8052           Personne    216
12      8283               Code  10022
13      9084               Code   7741
14      9084           Personne     17

z) Save the new dataframe in csv. Don't save your dataframe's index.

In [ ]:
df_processed_dim_name.to_csv("order_data_processed.csv", index=False)